resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+     - The set of test sites is the union of the test sites on either path.
+   - `a3 = a2`
+     - A variable is definitely assigned if it is definitely assigned in the
+       model of the finally block (note that the finally block is analyzed using
+       the join of the model from before the try block and after the try block,
+       and so assignments that may have occurred in the try block are already
+       modelled here).
+   - `u3 = u2`
+     - A variable is definitely unassigned if it is definitely unassigned in the
+       model of the finally block (note that the finally block is analyzed using
+       the join of the model from before the try block and after the try block,
+       and so the absence of any assignments that may have occurred in the try
+       block is already modelled here).
+   - `c3 = c2`
+     - A variable is captured if it is captured in the model of the finally
+       block (note that the finally block is analyzed using the join of the
+       model from before the try block and after the try block, and so any
+       captures from the try block are already modelled here).
+
+- `restrict(MB, MF, N)`, where `MB` and `MF` are flow models and `N` is a set of
+  variables assigned in the finally clause, models the flow of information
+  through a try/finally statement, and is defined as follows:
+
+  - We define `restrict(MB, MF, N)` to be `M3 = FlowModel(r3, VI3)` where:
+    - `MB = FlowModel(rb, VIB)`
+    - `MF = FlowModel(rf, VIF))`
+    - `pop(rb) = pop(rf) = r0` for some `r0`
+    - `r3` is `push(r0, top(rb) && top(rf))`
+    - `b` is true if `v` is in `N` and otherwise false
+    - `VI3` is the map which maps each variable `v` in the domain of `VIB` and
+      `VIF` to `restrictV(VIB(v), VIF(v), b)`.
+
+- `unreachable(M)` represents the model corresponding to a program location
+  which is unreachable, but is otherwise modeled by flow model `M = FlowModel(r,
+  VI)`, and is defined as `FlowModel(push(pop(r), false), VI)`
+
+- `demoteVariables(M, S)` represents the flow model derived from `M` in which
+  all variables in `S` have been demoted to their declared types.  It is defined
+  as `FlowModel(r, VI1)` where `M` is `FlowModel(r, VI0)` and `VI1` is the map
+  such that:
+    - `VI0` maps `v` to `VM0 = VariableModel(d0, p0, s0, a0, u0, c0)`
+    - If `S` contains `v` then `VI1` maps `v` to `VariableModel(d0, [], s0, a0, u0, c0)`
+    - Otherwise `VI1` maps `v` to `VM0`
+
+
+### Promotion
+
+Promotion policy is defined by the following operations on flow models.
+
+We say that the **current type** of a variable `x` in variable model `VM` is `S` where:
+  - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+  - `promoted = S::l` or (`promoted = []` and `declared = S`)
+
+Policy:
+  - We say that at type `T` is a type of interest for a variable `x` in a set of
+    tested types `tested` if `tested` contains a type `S` such that `T` is `S`,
+    or `T` is **NonNull(`S`)**.
```

Comment:
The "or `T` is **NonNull(`S`)**" part of this is not currently implemented.  Tracking in https://github.com/dart-lang/sdk/issues/40011.